### PR TITLE
MAINT: Drop convert_objects from NDFrame

### PR DESCRIPTION
**Pull request by gfyoung**
Deprecated since 0.17.0

xref #11173

### CODE REVIEW

File: pandas/core/internals.py [hunk 1]
```diff
@@ -1845,12 +1845,13 @@ def is_bool(self):
         """
         return lib.is_bool_array(self.values.ravel())
 
-    # TODO: Refactor when convert_objects is removed since there will be 1 path
     def convert(self, *args, **kwargs):
-        """ attempt to coerce any object types to better types return a copy of
-        the block (if copy = True) by definition we ARE an ObjectBlock!!!!!
+        """
+        Attempt to coerce any object types to more specific data types.
```
Comment (jreback) [inline pandas/core/internals.py]: there's a whole bunch of code here that i think u can ditch now that was for compat with convert_objects

Comment (gfyoung) [inline pandas/core/internals.py]: What do you mean by compat with `convert_objects`?  How so?

Comment (jreback) [inline pandas/core/internals.py]: most of this code iirc was to support multiple convert args at the same time (via convert_objects) which is not needed by convert - look at the blame and u shall see

Comment (gfyoung) [inline pandas/core/internals.py]: I don't really follow you here.  `convert_objects` was nowhere in the implementation of `convert`, and it seems like this function is used as a pathway to the replacement for the `object` conversion functions (e.g. `to_timedelta`)

File: pandas/core/internals.py [hunk 2]
```diff
@@ -1845,12 +1845,13 @@ def is_bool(self):
         """
         return lib.is_bool_array(self.values.ravel())
 
-    # TODO: Refactor when convert_objects is removed since there will be 1 path
```
Comment (jreback) [inline pandas/core/internals.py]: look and see when this comment was put in

Comment (jreback) [inline pandas/core/internals.py]: all of the fn_kwargs stuff can go back

Comment (jreback) [inline pandas/core/internals.py]: be removed i mean

Comment (gfyoung) [inline pandas/core/internals.py]: Yeah, I did.  #11173 - I'm not sure I follow what the logic was back then.

Comment (gfyoung) [inline pandas/core/internals.py]: Remove `fn_kwargs`?  `convert` appears to be just a wrapper around all of the object conversion functions as I mentioned above.  Doesn't that allow us to pass in keyword arguments specific to each of those functions?

**Comment by jorisvandenbossche** [comment]
Can we hold off on this one until we decide on https://github.com/pandas-dev/pandas/issues/11221 ?

(@gfyoung, sorry, when I saw you were doing all the deprecations, I wanted to raise this before you did this PR, but forgot it)

**Comment by jreback** [comment]
@jorisvandenbossche you would have to make a clear case for a useful conversion function
lots of arguments against 

not sure it's worth the name either

i would suggest a recipe or expand to_* to a) 2-d b) add a soft option

**Comment by jorisvandenbossche** [comment]
>  you would have to make a clear case

See my comment in https://github.com/pandas-dev/pandas/issues/11221#issuecomment-145072165 (at that time I was convinced of its usefulness, would have to dig in again for the details), and some of the other reactions. 
But lets keep that discussion there.

**Comment by jreback** [comment]
@jorisvandenbossche i saw your comment 

reviving a bad api is not going to happen 
you are welcome to propose a new one that has a new name
but that is orthogonal to this issue

**Comment by jreback** [comment]
@gfyoung what I mean is that [_possibly_convert_objects](https://github.com/pandas-dev/pandas/blob/master/pandas/types/cast.py#L567) should just become part of ``_soft_convert_objects`` (same file), prob rename -> ``_maybe_convert_objects``. This would be much easier once ``convert_objects`` is removed.

This *new* function is pretty much what @jorisvandenbossche is looking for an potentitally could be exposed. It allows soft & hard conversions.

**Comment by chris-b1** [comment]
I think it would be worth exposing whatever the new soft convert api is 0.20 (I haven't looked at it in detail), referencing it in the `convert_objects` depr message, then deferring `convert_objects` to the next version, if possible.

I say this because I know there are people (for example, me) who have ignored the `convert_objects` depr message in a couple cases, in particular working with data where you don't necessarily know the columns.  Real instance:

```python
df = pd.read_html(source)[0]  # poorly formatted table, everything inferred to object
                                                # exact columns can vary

df.columns = df.loc[0, :]
df = df.drop(0).dropna()

df = df.convert_objects()
```

Looking at this again, I realize`df.apply(lambda x: pd.to_numeric(x, errors='ignore'))` would also work fine in this case, but that wasn't immediately obvious, and I'm not sure we've done enough handholding (for lack of a better term) to help people transition.

**Comment by jorisvandenbossche** [comment]
@chris-b1 Can you comment that in https://github.com/pandas-dev/pandas/issues/11221 ?

**Comment by gfyoung** [comment]
@jorisvandenbossche @chris-b1 : You both raise some valid arguments regarding the keeping of `convert_objects`, so then I guess we have to back to @bashtage and ask it was deprecated in the first place :smile:.  I am perfectly fine with holding with regards to removal, though there were some minor doc fixes that I would still like to push through at minimum.  How does that sound?

**Comment by jreback** [comment]
@gfyoung 

it was deprecated because of the magic "if it has 1 valid value convert it, but if nothing converts then revert it" (maybe should make a song!)

So it becomes data dependent, not a great thing for a default. Sure it could be an *option* to the user which would be fine.

This further was exacerbated because the conversion would first try to convert to datetime, then if that failed to timdelta, then to numeric. Even more magic here.

So certainly we could add this back, but it would have to be an explict user decision.

**Comment by jreback** [comment]
this is basically what it does on top of the current functions: https://github.com/pandas-dev/pandas/issues/11221#issuecomment-246120793

**Comment by bashtage** [comment]
My opinion of `convert_objects` hasn't changed.  As @jreback said, it has a lot of *magic* and the precise behavior in all situations could only be described by the actual code being executed. This does not strike me as good design. 

The real issue here is that the well-defined rules for converting objects have already failed by the time that `convert_objects` was usually called (e.g., text data was imported from `read_`* and `object` was returned).  Given that there are not nearly lossless choices available, it seems very difficult to arrive at a reasonable set of rules that don't seem arbitrary.

If there was a need for a `convert_objects`-like function, then I think it would need to start from scratch by first describing a concrete set of actions that are more aggressive than the default converters but not so as to destroy the utility of this function.

**Comment by gfyoung** [comment]
@jreback : `_possibly_convert_objects` is in `cast.py`, which is not affected by this removal.  The functions which had the `TODO` are in `core/internals.py`  I think your proposed refactoring is beyond the scope of this removal, though it can implications on how to rebuild the API afterwards.

**Comment by jreback** [comment]
@gfyoung btw, no reason for you to keep rebasing this, unless something is actually changing.

**Comment by gfyoung** [comment]
Yeah, sorry, I have been making changes locally, but I haven't pushed things yet.

**Comment by jreback** [comment]
np by all means if u r changing things

**Comment by gfyoung** [comment]
@jreback : Is there anything else that needs to be changed here?  Even your proposed <a href="https://github.com/pandas-dev/pandas/pull/15757#issuecomment-288066093">diff</a> has already been incorporated into `master` already it seems

**Comment by gfyoung** [comment]
@jreback : Is there anything else that needs to be changed here?  Even your proposed <a href="https://github.com/pandas-dev/pandas/pull/15757#issuecomment-288066093">diff</a> has already been incorporated into `master` already it seems.

**Comment by jreback** [comment]
i think this is fine

even IF we want to have a future conversion function could always bring back the name 

i am
+1 in merging

**Comment by jorisvandenbossche** [comment]
I am -1 on merging this before a resolution on https://github.com/pandas-dev/pandas/issues/11221

**Comment by gfyoung** [comment]
@jorisvandenbossche : Could you clarify as to what a resolution would entail in #11221?  Also, I think @jreback had some questions that have yet to be answered about it.

**Comment by jreback** [comment]
@gfyoung no need to rebase this until we decide :>

**Comment by jreback** [comment]
i am in favor of merging this. No action has happened on the replacement since 0.20.0. If it is needed, someone can open an issue and/or implement this. If really really needed (which IMHO is not necessary anyhow).

**Comment by gfyoung** [comment]
@jreback : Don't forget that #11221 (which you opened regarding `convert_objects`) is still active.  So we also redirect people to re-open the discussion at a later point.

**Comment by gfyoung** [comment]
Also, if you do want to merge this in earnest, let me know, and I'll turn on testing to get the green light.

**Comment by jreback** [comment]
@jorisvandenbossche I don't think waiting on this is healthy. We haven't had any activity w.r.t. #11221. if someone really needs this then a new API can be designed. 

@chris-b1

**Comment by gfyoung** [comment]
@jreback : I might point out that you had mentioned moving forward on this about a month ago <a href="https://github.com/pandas-dev/pandas/pull/15757#issuecomment-306157021">here</a> with no response.  I am also in favor with merging this AND closing the related issue, which can always be revived in the future by anyone who's interested.

**Comment by jreback** [comment]
yes, let's close this and proceed as indicated in #11221, create ``infer_object`` and change the deprecation message on convert_objects (and schedule for next release). as indicated https://github.com/pandas-dev/pandas/issues/11221#issuecomment-314934018

**Pull request closed without merging**